In [ ]:
!pip install keras_tuner

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 176.1/176.1 kB 3.2 MB/s eta 0:00:00


In [ ]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf
import keras_tuner as kt

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

## Optimization Attempt 1: FAILED (Accuracy: 0.726)
## Steps Taken:
#### 1) Droped EIN, Name, & Special Considerations
#### 2) Changed application_type bin threshold from <500 to <100
#### 3) Changed classification bin threshold from <1000 to <500

In [ ]:
# Look at value counts; determine if necessary
application_df["SPECIAL_CONSIDERATIONS"].value_counts()

N    34272
Y       27
Name: SPECIAL_CONSIDERATIONS, dtype: int64

In [ ]:
# Drop columns, 'EIN', 'NAME', 'SPECIAL_CONSIDERATIONS'.
application_df = application_df.drop(["EIN", "NAME", "SPECIAL_CONSIDERATIONS"], axis=1)
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT   ASK_AMT  IS_SUCCESSFUL  
0       Association       1              0      5000              1  
1      Co-operative       1         1-9999    108590              1  
2       Association       1              0      5000              0  
3             Trust       1    10000-24999      6692              1  
4             Trust       1  100000-499999    142590              1  
...             ...     ...            ...       ...            ...  
34294   Association       1              0      5000              0  
34295   Association       1              0      5000              0  
34296   Association       1              0      5000              0  
34297   Association       1              0      5000              1  
34298  Co-operative       1          1M-5M  36500179              0  

[34299 rows x 9 columns]

In [ ]:
#Find value counts of application types to re-bin
application_df["APPLICATION_TYPE"].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [ ]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
# Get the counts of each category in 'APPLICATION_TYPE'
app_type_counts = application_df.APPLICATION_TYPE.value_counts()

# Bin categories with count <500
application_types_to_replace = list(app_type_counts[app_type_counts < 100].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
T9         156
Other      120
Name: APPLICATION_TYPE, dtype: int64

In [ ]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE      10
AFFILIATION            6
CLASSIFICATION        71
USE_CASE               5
ORGANIZATION           4
STATUS                 2
INCOME_AMT             9
ASK_AMT             8747
IS_SUCCESSFUL          2
dtype: int64

In [ ]:
# Look at CLASSIFICATION value counts for binning
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [ ]:
# You may find it helpful to look at CLASSIFICATION value counts >1
# Get value counts
classification_counts = application_df['CLASSIFICATION'].value_counts()

# Filter
classification_counts = classification_counts[classification_counts > 1]

print(classification_counts)

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64


In [ ]:
# Change bins on classifications from <1000 to <500
# Set classifications with counts < 500 to "Other"
application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].where(application_df['CLASSIFICATION'].map(classification_counts) >= 500, 'Other')

# Verify the transformation
print(application_df['CLASSIFICATION'].value_counts())

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
Other     1484
C7000      777
Name: CLASSIFICATION, dtype: int64


In [ ]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df_dummies = pd.get_dummies(application_df)
application_df_dummies.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                       0   
1       1   108590              1                       0   
2       1     5000              0                       0   
3       1     6692              1                       0   
4       1   142590              1                       0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                     1                     0                    0   
1                     0                     0                    1   
2                     0                     0                    0   
3                     0                     0                    1   
4                     0                     0                    1   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                    0                    0                    0  ...   
1                    0                    0                    0  ...   
2                    0                    1                    0  ...   
3                    0                    0                    0  ...   
4                    0                    0                    0  ...   

   ORGANIZATION_Trust  INCOME_AMT_0  INCOME_AMT_1-9999  \
0                   0             1                  0   
1                   0             0                  1   
2                   0             1                  0   
3                   1             0                  0   
4                   1             0                  0   

   INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                       0                         0                   0   
1                       0                         0                   0   
2                       0                         0                   0   
3                       1                         0                   0   
4                       0                         1                   0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0                 0                       0                0   
1                 0                       0                0   
2                 0                       0                0   
3                 0                       0                0   
4                 0                       0                0   

   INCOME_AMT_5M-10M  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  

[5 rows x 44 columns]

In [ ]:
# Split our preprocessed data into our features and target arrays
# Separate the y variable, the labels
y = application_df_dummies['IS_SUCCESSFUL']

# Separate the X variable, the features
X = application_df_dummies.drop('IS_SUCCESSFUL', axis=1)

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [ ]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [ ]:
# Get the number of features from the training data
num_features = X_train_scaled.shape[1]

# Define the model - deep neural net
nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=6, activation="relu", input_dim=num_features))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=6, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 6)                 264       
                                                                 
 dense_1 (Dense)             (None, 6)                 42        
                                                                 
 dense_2 (Dense)             (None, 1)                 7         
                                                                 
Total params: 313
Trainable params: 313
Non-trainable params: 0
_________________________________________________________________


In [ ]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [ ]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 3s 2ms/step - loss: 0.6477 - accuracy: 0.6621
Epoch 2/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5866 - accuracy: 0.7176
Epoch 3/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5713 - accuracy: 0.7259
Epoch 4/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5643 - accuracy: 0.7266
Epoch 5/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5604 - accuracy: 0.7268
Epoch 6/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5578 - accuracy: 0.7288
Epoch 7/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5561 - accuracy: 0.7293
Epoch 8/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5552 - accuracy: 0.7304
Epoch 9/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5543 - accuracy: 0.7283
Epoch 10/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5535 - accuracy: 0.7287

In [ ]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5558 - accuracy: 0.7264 - 1s/epoch - 5ms/step
Loss: 0.5557546019554138, Accuracy: 0.7264139652252197


## Optimization Attempt 2: FAILED (Accuracy: 0.7264)
## Steps Taken:
#### 1) Droped EIN, Name, & Special Considerations
#### 2) Changed application_type bin threshold from <500 to <100
#### 3) Changed classification bin threshold from <1000 to <500
#### 4) Changed activiation functions to relu, tanh, and sigmoid

In [ ]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
# Get the counts of each category in 'APPLICATION_TYPE'
app_type_counts = application_df.APPLICATION_TYPE.value_counts()

# Bin categories with count <500
application_types_to_replace = list(app_type_counts[app_type_counts < 100].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
T9         156
Other      120
Name: APPLICATION_TYPE, dtype: int64

In [ ]:
# Change bins on classifications from <1000 to <500
# Set classifications with counts < 500 to "Other"
application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].where(application_df['CLASSIFICATION'].map(classification_counts) >= 500, 'Other')

# Verify the transformation
print(application_df['CLASSIFICATION'].value_counts())

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
Other     1484
C7000      777
Name: CLASSIFICATION, dtype: int64


In [ ]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df_dummies = pd.get_dummies(application_df)
application_df_dummies.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                       0   
1       1   108590              1                       0   
2       1     5000              0                       0   
3       1     6692              1                       0   
4       1   142590              1                       0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                     1                     0                    0   
1                     0                     0                    1   
2                     0                     0                    0   
3                     0                     0                    1   
4                     0                     0                    1   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                    0                    0                    0  ...   
1                    0                    0                    0  ...   
2                    0                    1                    0  ...   
3                    0                    0                    0  ...   
4                    0                    0                    0  ...   

   ORGANIZATION_Trust  INCOME_AMT_0  INCOME_AMT_1-9999  \
0                   0             1                  0   
1                   0             0                  1   
2                   0             1                  0   
3                   1             0                  0   
4                   1             0                  0   

   INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                       0                         0                   0   
1                       0                         0                   0   
2                       0                         0                   0   
3                       1                         0                   0   
4                       0                         1                   0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0                 0                       0                0   
1                 0                       0                0   
2                 0                       0                0   
3                 0                       0                0   
4                 0                       0                0   

   INCOME_AMT_5M-10M  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  

[5 rows x 44 columns]

In [ ]:
# Split our preprocessed data into our features and target arrays
# Separate the y variable, the labels
y = application_df_dummies['IS_SUCCESSFUL']

# Separate the X variable, the features
X = application_df_dummies.drop('IS_SUCCESSFUL', axis=1)

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [ ]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [ ]:
# Get the number of features from the training data
num_features = X_train_scaled.shape[1]

# Define the model - deep neural net
nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=6, activation="relu", input_dim=num_features))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=6, activation="tanh"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 6)                 264       
                                                                 
 dense_4 (Dense)             (None, 6)                 42        
                                                                 
 dense_5 (Dense)             (None, 1)                 7         
                                                                 
Total params: 313
Trainable params: 313
Non-trainable params: 0
_________________________________________________________________


In [ ]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [ ]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 3s 2ms/step - loss: 0.6123 - accuracy: 0.6956
Epoch 2/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5667 - accuracy: 0.7217
Epoch 3/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5599 - accuracy: 0.7252
Epoch 4/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5575 - accuracy: 0.7258
Epoch 5/50
804/804 [==============================] - 3s 4ms/step - loss: 0.5559 - accuracy: 0.7269
Epoch 6/50
804/804 [==============================] - 3s 4ms/step - loss: 0.5541 - accuracy: 0.7271
Epoch 7/50
804/804 [==============================] - 3s 4ms/step - loss: 0.5529 - accuracy: 0.7275
Epoch 8/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5521 - accuracy: 0.7276
Epoch 9/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5515 - accuracy: 0.7279
Epoch 10/50
804/804 [==============================] - 3s 4ms/step - loss: 0.5513 - accuracy: 0.7282

In [ ]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5552 - accuracy: 0.7264 - 1s/epoch - 4ms/step
Loss: 0.555242121219635, Accuracy: 0.7264139652252197


## Optimization Attempt 3: FAILED (Accuracy: 0.7296)
## Steps Taken:
#### 1) Droped EIN, Name, & Special Considerations
#### 2) Changed application_type bin threshold from <500 to <100
#### 3) Changed classification bin threshold from <1000 to <500
#### 4) Changed activiation functions to relu and sigmoid
#### 5) Added additional hidden layer, increased units from 6 to 10

In [ ]:
# Split our preprocessed data into our features and target arrays
# Separate the y variable, the labels
y = application_df_dummies['IS_SUCCESSFUL']

# Separate the X variable, the features
X = application_df_dummies.drop('IS_SUCCESSFUL', axis=1)

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [ ]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [ ]:
# Get the number of features from the training data
num_features = X_train_scaled.shape[1]

# Define the model - deep neural net
nn = tf.keras.models.Sequential()

# First hidden layer, more units
nn.add(tf.keras.layers.Dense(units=10, activation="relu", input_dim=num_features))

# Second hidden layer, more units
nn.add(tf.keras.layers.Dense(units=10, activation="relu"))

# Third hidden layer, more units
nn.add(tf.keras.layers.Dense(units=10, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 10)                440       
                                                                 
 dense_7 (Dense)             (None, 10)                110       
                                                                 
 dense_8 (Dense)             (None, 10)                110       
                                                                 
 dense_9 (Dense)             (None, 1)                 11        
                                                                 
Total params: 671
Trainable params: 671
Non-trainable params: 0
_________________________________________________________________


In [ ]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [ ]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 3s 2ms/step - loss: 0.5955 - accuracy: 0.6993
Epoch 2/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5584 - accuracy: 0.7276
Epoch 3/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5540 - accuracy: 0.7287
Epoch 4/50
804/804 [==============================] - 4s 5ms/step - loss: 0.5518 - accuracy: 0.7312
Epoch 5/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5504 - accuracy: 0.7304
Epoch 6/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5494 - accuracy: 0.7308
Epoch 7/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5485 - accuracy: 0.7315
Epoch 8/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5479 - accuracy: 0.7323
Epoch 9/50
804/804 [==============================] - 3s 4ms/step - loss: 0.5477 - accuracy: 0.7324
Epoch 10/50
804/804 [==============================] - 4s 5ms/step - loss: 0.5474 - accuracy: 0.7328

In [ ]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5517 - accuracy: 0.7297 - 535ms/epoch - 2ms/step
Loss: 0.5517356991767883, Accuracy: 0.72967928647995


In [ ]:
# Export optimized model to HDF5 file
nn.save("AlphabetSoupCharity_Optimization.h5")